Feature/grading instructions/modify data model for SGI

HanyaElhashemy opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Modify the database model in order to extend the new structured grading instructions feature in Artemis
### Description
<!-- Describe your changes in detail -->
Added the new table "grading_instruction" to the database
The table has the following columns:
- credit: is the score students get if this criteria is applicable
- level: is the grading scale, for example “poor”, “average” or “good”
- instruction_description: is the description of the criteria that helps the tutor to choose whether the criteria can be applied to the given submission
- feedback: is an explanation of the mistake students did and a description of how this mistake could be improved in the future
- usage_count: here the instructor specifies how often points for this criteria should be calculated if the criteria is applicable more than once for the same submission (e.g. to prevent consequential errors)

To enable group instructions of a specific learning objective together we also introduce the new table grading_criteria. Grading instruction(s) could be either part of a criterion or stand-alone.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Test if all previous functionality still works as expected

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

Analysis object model
![image](https://user-images.githubusercontent.com/50635005/74224780-c1a99a80-4cb9-11ea-9fca-9bdc9e43129a.png)

lovaiible commented on src/main/resources/config/liquibase/changelog/20191210131150_changelog.xml:
```diff
@@ -0,0 +1,17 @@
+                <constraints primaryKey="true"/>
+            </column>
+            <column name="credits" type="DOUBLE"/>
+            <column name="level" type="VARCHAR(255)"/>
```
I don't know the concept behind the `level` column, but we can consider it being an ENUM right? Or does every Instructor/TA define their own levels?

HanyaElhashemy replied:
Good point! I think we need to discuss this further and look at the advantages and disadvantages of giving instructors the freedom to define their own grading scale.

beatrisbu replied:
We can discuss it on Tuesday but I think the way you meant it to be is perfectly okay - “poor”, “average”, “good” or another more detailed, but still predefined, list makes more sense in my opinion as otherwise it could get really inconsistent

HanyaElhashemy commented on src/main/java/de/tum/in/www1/artemis/domain/GradingInstruction.java:
```diff
@@ -0,0 +1,156 @@
+    @Column(name = "level")
+    private String level;
+
+    @Enumerated(EnumType.STRING)
```
You are right, this does not make sense ^^'

HanyaElhashemy commented on src/main/java/de/tum/in/www1/artemis/domain/GradingInstruction.java:
```diff
@@ -0,0 +1,156 @@
+    private Double credits;
+
+    @Column(name = "level")
+    private String level;
```
I got the naming from this paper describing the parts of a grading rubric:
http://bsuenglish101.pbworks.com/f/rubricshandout.pdf
Here the grading scale was named the "level of performance", but I am not sure which naming is more intuitive for developers tbh.

HanyaElhashemy replied:
As you agreed on the naming I changed it, thanks for the suggestion. It was also very useful to rename it as "level" is used somewhere else to describe the difficulty level of exercises.

HanyaElhashemy commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,151 @@
+ */
+@RestController
+@RequestMapping({ "/api", "/api_basic" })
+@PreAuthorize("hasRole('ADMIN')")
```
I think so since all methods are executed by Admin and Instructor. So maybe I should modify it and add the Instructor as well.

HanyaElhashemy commented:
> Tested following functionalities:
> 
> * create Text/Modeling/File Upoad Exercise with grading instructions
> * Grading instructions are shown to the tutor and have to be accepted before first assessment
> * Grading instructions are displayed correctly on the assessment page and can be minimized
> 
> Not sure if that was all of the "previous functionality" related to this PR. Maybe you can update the "Steps for Testing", that would probably make it clearer, at least for me. :)

You are right, actually, I did not modify any already existing functionality so testing it should not be relevant. Code review is more important in this case! It would be nice if you check the code and the newly added database and see if the API is configured correctly. Thanks!

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    /**
+     * GET /exercises/:id : get the "id" exercise.
```
JavaDoc comment does not match what the method actually does.

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,151 @@
+     * @param exerciseId the id of the exercise to retrieve
+     * @return the ResponseEntity with status 200 (OK) and with body the exercise, or with status 404 (Not Found)
+     */
+    @GetMapping("/grading-instruction/{exerciseId}")
```
The route name is not conform with our REST naming conventions (see [here](https://confluence.ase.in.tum.de/display/ArTEMiS/Server)). Specifically, it violates these points:

1. plural of entity name should normally be used (`/grading-instructions`)
2. the placement of the `exerciseId` param is misleading since a call to e.g. `/grading-instruction/123` insinuates that `123` is the gradient instruction id (and not the exercise id).
3. the url structure should go from entity to subentity, meaning in this case `/exercises/{exerciseId}/gradient-instructions` would probably be a good choice

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,151 @@
+        if (!authCheckService.isAtLeastTeachingAssistantForExercise(exercise)) {
+            return forbidden();
+        }
+        List<GradingInstruction> gradingInstructions = gradingInstructionService.findAllForExercise(exercise);
```
Is this additional db query needed? Couldn't `exercise.getStructuredGradingInstructions()` be used instead since you already loaded the grading instructions eagerly with the exercise?

HanyaElhashemy replied:
Very Good point, thank you!

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,151 @@
+     * @return the ResponseEntity with status 201 (Created) and with body the new gradingInstruction, or with status 400 (Bad Request) if the gradingInstruction has already an ID
+     * @throws URISyntaxException if the Location URI syntax is incorrect
+     */
+    @PostMapping("/grading-instruction")
```
Should be plural.

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,151 @@
+        Exercise exercise = exerciseService.findOne(gradingInstruction.getExercise().getId());
+        if (exercise == null) {
+            return ResponseEntity.badRequest().headers(
+                    HeaderUtil.createFailureAlert(applicationName, true, ENTITY_NAME, "exerciseNotFound", "The exercise belonging to this grading instructor does not exist"))
```
```suggestion
                    HeaderUtil.createFailureAlert(applicationName, true, ENTITY_NAME, "exerciseNotFound", "The exercise belonging to this grading instruction does not exist"))
```

HanyaElhashemy replied:
I am impressed by how you find these mistakes ^^

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,151 @@
+
+        GradingInstruction gradingInstruction = gradingInstructionService.findOne(gradingInstructionId);
+        Exercise exercise = exerciseService.findOne(gradingInstruction.getExercise().getId());
+        if (Optional.ofNullable(gradingInstruction).isPresent()) {
```
What if this condition is false? At the moment the endpoint would still return ok as if deletion was successful. And wouldn't there be an issue in the previous line if `gradingInstruction` were null?

ungaralex replied:
This is not resolved. You didn't change anything here.

madwau commented on src/main/resources/config/liquibase/changelog/20191210131150_changelog.xml:
```diff
@@ -0,0 +1,17 @@
+            <column name="usage_count" type="INT"/>
+            <column name="exercise_id" type="BIGINT"/>
+        </createTable>
+        <addForeignKeyConstraint baseColumnNames="exercise_id" baseTableName="grading_instruction" constraintName="fk_grading_instruction_exercise_id" deferrable="false" initiallyDeferred="false" onDelete="NO ACTION" onUpdate="NO ACTION" referencedColumnNames="id" referencedTableName="exercise"/>
```
Shouldn't the grading instruction also be deleted if the exercise itself was deleted?

ungaralex commented on src/main/java/de/tum/in/www1/artemis/domain/GradingInstruction.java:
```diff
@@ -0,0 +1,155 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
```
Nevermind, this was a bad advice

HanyaElhashemy commented on src/main/java/de/tum/in/www1/artemis/domain/GradingInstruction.java:
```diff
@@ -0,0 +1,155 @@
+@Entity
+@Table(name = "grading_instruction")
+@Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+public class GradingInstruction implements Serializable {
```
I find StructuredGradingInstruction to be a bit long and I think the idea is to delete the string attribute in the future when my feature is available.

HanyaElhashemy commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,151 @@
+    private static final String ENTITY_NAME = "gradingInstruction";
+
+    @Value("${jhipster.clientApp.name}")
+    private String applicationName;
```
@ungaralex Could you please explain why I should rename it? I used the same naming as in all other resources ^^

ungaralex replied:
From what I've seen in other services/resources/etc., we decided to use the constant naming convention for our injected values. There are some exceptions (i.e. the application name) where this is not the case for some resource, but we should try to stick to the other naming scheme for new code.

HanyaElhashemy commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,151 @@
+     * @return the ResponseEntity with status 200 (OK) and with body the exercise, or with status 404 (Not Found)
+     */
+    @GetMapping("/grading-instruction/{exerciseId}")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
```
You are right! We do not need Security Annotation at Class level.

HanyaElhashemy commented on src/test/java/de/tum/in/www1/artemis/GradingInstructionTest.java:
```diff
@@ -0,0 +1,114 @@
+        long courseID = courseRepository.findAllActive().get(0).getId();
+        exercise = exerciseRepository.findByCourseId(courseID).get(0);
+
+        gradingInstructionSet = database.addGradingInstructionsToExercise(exercise).getStructuredGradingInstructions();
```
But instructions don't exist on their own, they are being directly added to the exercise. It is a composition relationship between the exercise and the instructions.

HanyaElhashemy commented:
@lovaiible @madwau @ungaralex Thank you for investing time in my PR and adding these valuable comments on my code, I appreciate it. I worked on your comments and resolved the remarks mentioned above.

HanyaElhashemy commented on src/main/java/de/tum/in/www1/artemis/domain/GradingInstruction.java:
```diff
@@ -0,0 +1,155 @@
+        return exercise;
+    }
+
+    public GradingInstruction result(Exercise exercise) {
```
You are right, the naming of the method was misleading ^^'

HanyaElhashemy commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,155 @@
+     * (Internal Server Error) if the gradingInstruction couldn't be updated
+     * @throws URISyntaxException if the Location URI syntax is incorrect
+     */
+    @PutMapping("/grading-instructions")
```
@lovaiible So as I understood from the REST Controller: Route naming conventions on confluence we should avoid specifying an id if possible and simplify the route. Did I maybe misunderstand it in this case? I also looked at the `courseResource` and took it as a reference.

lovaiible replied:
@HanyaElhashemy My bad then. I did not know about this naming convention. If it's recommended as such, then it's ok ;)

ungaralex commented on src/main/java/de/tum/in/www1/artemis/service/GradingInstructionService.java:
```diff
@@ -0,0 +1,76 @@
+     */
+
+    public GradingInstruction findOne(long gradingInstructionId) {
+        Optional<GradingInstruction> gradingInstruction = gradingInstructionRepository.findById(gradingInstructionId);
```
Nit
```suggestion
        return gradingInstructionRepository.findById(gradingInstructionId)
                .orElseThrow(() -> new EntityNotFoundException("Exercise with exerciseId " + gradingInstructionId + " does not exist!"));
```

ungaralex commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,145 @@
+     */
+    @GetMapping("/exercises/{exerciseId}/grading-instructions")
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<Set<GradingInstruction>> getGradingInstructionsByExerciseId(@PathVariable Long exerciseId) {
```
```suggestion
    public ResponseEntity<Set<GradingInstruction>> getGradingInstructionsByExerciseId(@PathVariable long exerciseId) {
```

ungaralex commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,145 @@
+     */
+    @DeleteMapping("/grading-instructions/{gradingInstructionId}")
+    @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<Void> deleteGradingInstruction(@PathVariable Long gradingInstructionId) {
```
```suggestion
    public ResponseEntity<Void> deleteGradingInstruction(@PathVariable long gradingInstructionId) {
```

ungaralex commented on src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java:
```diff
@@ -423,6 +432,16 @@ public Result addResultToSubmission(Submission submission) {
         return result;
     }
 
+    public Set<GradingInstruction> addGradingInstructionsToExercise(Exercise exercise) {
```
Maybe add the amount of instructions to add as a parameter

HanyaElhashemy replied:
Could you please explain why I should add the number of instructions?

ungaralex replied:
So that developers that read through code, which calls this method, immediately know how many instructions get created and so that you can be flexible in future tests (maybe you will have some tests that require only one instruction, maybe more, etc.)

ungaralex commented on src/main/java/de/tum/in/www1/artemis/domain/GradingInstruction.java:
```diff
@@ -0,0 +1,155 @@
+    private double credits;
+
+    @Column(name = "grading_scale")
+    private String gradingScale;
```
Is this nullable? Moreover, which of the attributes should be nullable, and which should not?

HanyaElhashemy replied:
I did not add such limitations yet. I think I need to discuss these design decisions with Stephan first.

ungaralex commented on src/main/java/de/tum/in/www1/artemis/domain/GradingInstruction.java:
```diff
@@ -0,0 +1,155 @@
+    private String feedback;
+
+    @Column(name = "usage_count")
+    private int usageCount;
```
From your explanation, isn't this more an `applicationLimit` than a count?

HanyaElhashemy replied:
So "usage_count" is the number instructors enter to indicate when the mistake should be counted as a subsequential mistake.
for example: 0 means every time this mistake is made we should **use** the grading instruction and subtract points.
x means that we should subtract points x times if the mistake occurred more than x times aka we should **use** the grading instruction maximum x times only, that is why I named it usage_count.

ungaralex replied:
Ok, so in this case `Integer` would make more sense in this case imo since `null` would be the explicit value if you want to just apply the instruction every time. 
I think you should also add what you just wrote in your comment in the JavaDoc since this is a behavior one can't deduce from just looking at the variable.

ungaralex commented on src/main/java/de/tum/in/www1/artemis/domain/GradingInstruction.java:
```diff
@@ -0,0 +1,155 @@
+@Entity
+@Table(name = "grading_instruction")
+@Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+public class GradingInstruction implements Serializable {
```
You could add some JavaDoc to the attributes with a short explanation (e.g. the one you are using in your proposal). Having this is always really nice for new developers.

ungaralex commented on src/main/java/de/tum/in/www1/artemis/web/rest/GradingInstructionResource.java:
```diff
@@ -0,0 +1,145 @@
+     * @param exerciseId the id of the exercise to retrieve its grading instructions
+     * @return the ResponseEntity with status 200 (OK) and with body the exercise, or with status 404 (Not Found)
+     */
+    @GetMapping("/exercises/{exerciseId}/grading-instructions")
```
Please refactor these endpoint Strings into a public constant subclass `Endpoints`. We already did this for programming exercises (`ProgrammingExerciseResource`) and a part of the `ParticipationResource`. This helps keeping the Strings between tests and resources (and also within the resource itself) consistent.

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:
```diff
@@ -104,6 +104,11 @@
     @JsonView(QuizView.Before.class)
     private Course course;
 
+    @OneToMany(mappedBy = "exercise", fetch = FetchType.EAGER, cascade = CascadeType.REMOVE)
```
please do not use FetchType.EAGER

krusche replied:
@OneToMany(mappedBy = "exercise", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderColumn
    @JsonIgnoreProperties(value = "exercise", allowSetters = true)
    @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
    private List< GradingInstruction > structuredGradingInstructions = new ArrayList<>();

HanyaElhashemy commented:
@krusche @ungaralex @madwau @lovaiible I modified the data-model, by introducing "criteria" as a parent object of a list of grading instructions. Now instructions describing the same learning objective could be grouped together and mapped to a criterion with a criterion title. I also removed the rest calls for grading instructions as criteria with its instructions get directly created when creating an exercise.
Since the model changed a bit I would need your help again in reviewing this PR (I apologize for that) so would you be so kind and have a look at it. Thanks!

krusche commented on src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:
```diff
@@ -336,4 +336,22 @@ public void calculateNrOfOpenComplaints(Exercise exercise) {
+     *
+     * @param exercise Exercise with criteria
+     */
+    public void mapExerciseToCriteria(Exercise exercise) {
```
I think this is unnecessary if you use the right annotations.

krusche replied:
@JsonIgnoreProperties(value = "result", allowSetters = true)

krusche replied:
if at all, I would rename this to reconnectCriteriaWithExercise

krusche commented on src/test/java/de/tum/in/www1/artemis/ModelingExerciseIntegrationTest.java:
```diff
@@ -103,7 +114,10 @@ public void testGetModelingExerciseStatistics_asTA() throws Exception {
     public void testCreateModelingExercise_asInstructor() throws Exception {
         ModelingExercise modelingExercise = modelingExerciseUtilService.createModelingExercise(classExercise.getCourse().getId());
-        request.post("/api/modeling-exercises", modelingExercise, HttpStatus.CREATED);
+        ModelingExercise receivedModelingExercise = request.postWithResponseBody("/api/modeling-exercises", modelingExercise, ModelingExercise.class, HttpStatus.CREATED);
```
also add a test, where you invoke PUT after POST and where you have changed some of the criteria (i.e. add a new one, change an existing one, delete an existing one) and the instructions

krusche commented on src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java:
```diff
@@ -463,6 +479,20 @@ public Result addResultToSubmission(Submission submission) {
     }
 
+    public List<GradingCriteria> addGradingInstructionsToExercise(Exercise exercise) {
+        GradingCriteria emptyCriteria = ModelFactory.generateGradingCriteria(null);
```
GradingCriterion

krusche commented on src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java:
```diff
@@ -463,6 +479,20 @@ public Result addResultToSubmission(Submission submission) {
+        List<GradingInstruction> instructionWithNoCriteria = ModelFactory.generateGradingInstructions(emptyCriteria, 1);
+        emptyCriteria.setStructuredGradingInstructions(instructionWithNoCriteria);
+        GradingCriteria testCriteria = ModelFactory.generateGradingCriteria("test title");
+        List<GradingInstruction> instructions = ModelFactory.generateGradingInstructions(testCriteria, 3);
```
`testCriteria.setExercise(exercise)`

HanyaElhashemy commented:
@krusche I applied your feedback and was able to remove the `mapExerciseToCriteria` Method. Could you please review it again?

lovaiible commented on src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:
```diff
@@ -796,4 +802,39 @@ public Boolean getPresentationScoreEnabled() {
+    public Exercise gradingCriteria(List<GradingCriterion> gradingCriteria) {
+        this.gradingCriteria = gradingCriteria;
+        if (gradingCriteria != null) {
+            this.gradingCriteria.forEach(gradingCriterion -> {
```
The code here and in line `833` are the same. Maybe you can move it to a private method? :)

lovaiible commented on src/test/java/de/tum/in/www1/artemis/FileUploadExerciseIntegrationTest.java:
```diff
@@ -62,11 +67,19 @@ public void createFileUploadExercise() throws Exception {
         assertThat(receivedFileUploadExercise.getId()).isNotNull();
         assertThat(receivedFileUploadExercise.getFilePattern()).isEqualTo(filePattern.toLowerCase().replaceAll("\\s+", ""));
+
+        assertThat(receivedFileUploadExercise.getGradingCriteria().get(0).getTitle()).isEqualTo(null);
```
Just a nitpick, but I saw that you have this Assertion (and the following ones) in the method `testGetModelingExercise_asTA` as well. You could actually move it into your method `addGradingInstructionsToExercise`, so you don't have to add the Assertion every time you call `addGradingInstructionsToExercise`

madwau commented on src/main/java/de/tum/in/www1/artemis/domain/GradingInstruction.java:
```diff
@@ -0,0 +1,159 @@
+
+    // describes the level of performance (e.g. "poor" or "average" )
+    @Column(name = "grading_scale")
+    private String gradingScale;
```
Could this be an enum?

HanyaElhashemy replied:
I discussed this with Stephan and we decided to leave it for now and maybe change it later to enum if it makes sense.

jpbernius commented on src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:
```diff
@@ -796,4 +802,39 @@ public Boolean getPresentationScoreEnabled() {
+            this.gradingCriteria.forEach(gradingCriterion -> {
+                gradingCriterion.setExercise(this);
+            });
+        }
```
```java
setGradingCriteria(gradingCriteria)
```

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/TextExerciseResource.java:
```diff
@@ -49,6 +47,8 @@
 
     private final TextExerciseRepository textExerciseRepository;
 
+    private final GradingCriterionRepository gradingCriterionRepository;
```
This is not used.

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/ExerciseResource.java:
```diff
@@ -326,9 +326,9 @@ else if (exercise instanceof ProgrammingExercise) {
         if (exercise != null) {
             List<StudentParticipation> participations = participationService.findByExerciseIdAndStudentIdWithEagerResultsAndSubmissions(exercise.getId(), user.getId());
-
+            List<GradingCriterion> gradingCriteria = gradingCriterionService.findByExerciseIdWithEagerGradingCriteria(exercise.getId());
```
I am wondering if this should be fetched in the same call as the exercise above. This would save an extra database call.

HanyaElhashemy replied:
You're right! They are already fetched above when the exercise is fetched so no need for this database call. I will remove it!

HanyaElhashemy replied:
I am not sure if this is possible as the criteria are not set by the time the exercise gets fetched above.

jpbernius replied:
Well you would need to use something like the `findByIdWithEagerGradingCriteria` method you defined in this PR instead of the service call to just get the Exercise.

HanyaElhashemy replied:
Ah, I get your point now and changed it accordingly! Thanks for mentioning it.

maxr96 commented on src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:
```diff
@@ -336,4 +336,5 @@ public void calculateNrOfOpenComplaints(Exercise exercise) {
         exercise.setNumberOfOpenMoreFeedbackRequests(numberOfMoreFeedbackRequests - numberOfMoreFeedbackComplaintResponses);
         exercise.setNumberOfMoreFeedbackRequests(numberOfMoreFeedbackRequests);
     }
+
```
Please remove blank line, since you didn't change anything in this file

maxr96 commented on src/main/java/de/tum/in/www1/artemis/repository/GradingCriterionRepository.java:
```diff
@@ -0,0 +1,22 @@
+    List<GradingCriterion> findByExerciseId(long exerciseId);
+
+    @Query("select distinct criterion from GradingCriterion criterion left join fetch criterion.structuredGradingInstructions where criterion.exercise.id = :#{#exerciseId}")
+    List<GradingCriterion> findByExerciseIdWithEagerGradingCriteria(@Param("exerciseId") Long exerciseId);
```
Can you use here `long` as well to stay consistent?

maxr96 commented on src/main/java/de/tum/in/www1/artemis/service/GradingCriterionService.java:
```diff
@@ -0,0 +1,49 @@
+     * @param exerciseId the id of exercise
+     * @return the list of criteria belonging to exercise
+     */
+    public List<GradingCriterion> findByExerciseIdWithEagerGradingCriteria(Long exerciseId) {
```
And here also `long`

maxr96 commented on src/main/java/de/tum/in/www1/artemis/domain/GradingInstruction.java:
```diff
@@ -0,0 +1,159 @@
+        return this;
+    }
+
+    public void setCredits(Double credits) {
```
nit-pick: here should be `double` instead of `Double`

maxr96 commented on src/main/java/de/tum/in/www1/artemis/domain/GradingInstruction.java:
```diff
@@ -0,0 +1,159 @@
+
+    // the score students get if this grading instruction is applicable
+    @Column(name = "credits")
+    private double credits;
```
Can the credits be null here?

HanyaElhashemy replied:
no
